File: hackerrank/journey_to_mars.cpp
#include <cmath>
#include <cstdio>
#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

long int integer_power(long int N, long int M) {
    long int result = 1;
    if(M>100)
        return pow(N, M);
    for(long int i=0; i<M; i++)
        result *= N;
    return result;
}

long int last_k_digits(long int N, long int TENK) {
    long int result = 1;
    if(N == 1)
        return result;
    else if (N % 2 == 0) {
        result *= last_k_digits(N/2, TENK);
        result %= TENK;
        result *= last_k_digits(N/2, TENK);
        result %= TENK;
    }
    else {
        long int result = 1;
        result *= last_k_digits((N-1)/2, TENK);
        result %= TENK;
        result *= last_k_digits((N-1)/2, TENK);
        result %= TENK;
        result *= 2;
        result %= TENK;
    }
    return result;
}

int main(int argc, char ** argv){
    int T;
    cin >> T;
    for(int i=0; i<T; i++){
        long int N, K;
        cin >> N >> K;
        int mval = K * log(10)/log(2);
        long int tenk = integer_power(10, K);
        long int lowest_k = last_k_digits(N-1, tenk);
        long double n1log102 = (N-1) * log(2)/log(10);
        long double power = n1log102 - int(n1log102) + K - 1;
        long int highest_k = pow(10, power);
        cout << (highest_k+lowest_k) << endl;
    }
}

File: hackerrank/journey_to_mars.py
import math
for _ in range(int(input())):
    N, K = [int(x) for x in input().split()]
    mval = int(K * math.log(10)/math.log(2))
    if N-1 <= mval:
        lowest_k = 2**(N-1) % 10**K
    else:
        pow2_mval = 2**mval
        lowest_k = pow2_mval % 10**K
        pow2 = N-1-mval
        while pow2 > mval:
            lowest_k *= pow2_mval
            lowest_k %= 10**K
            pow2 -= mval
        lowest_k *= 2**pow2
        lowest_k %= 10**K
        
    n1log102 = (N-1) * math.log10(2)
    power = n1log102 - int(n1log102) + K - 1
    highest_k = int(10**power)
    print(highest_k+lowest_k)
